sdk/lib/dmilib/dmilib.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * PROJECT:     ReactOS DMI/SMBIOS Library
+ * LICENSE:     GPL - See COPYING in the top level directory
+ * FILE:        dmilib.c
+ * PURPOSE:     SMBIOS table parsing functions
+ * PROGRAMMERS: Timo Kreuzer (timo.kreuzer@reactos.org)
+ * REFERENCES:  - http://www.dmtf.org/sites/default/files/standards/documents/DSP0134_3.0.0.pdf
+ *              -
+ */
+
+#include "precomp.h"
+
+static
+PCHAR
+GetDmiString(
+    _In_ PDMI_HEADER Header,
+    _In_ ULONG FieldOffset)
+{
+    ULONG StringIndex;
+    PCHAR String;
+
+    StringIndex = ((PUCHAR)Header)[FieldOffset];
+    if (StringIndex == 0)
+    {
+        return NULL;
+    }
+
+    String = (PCHAR)Header + Header->Length;
+
+    while (--StringIndex != 0)
+    {
+        while (*String != 0)
+            String++;
+
+        String++;
+    }
+
+    return String;
+}
+
+VOID
+ParseSMBiosTables(
+    _In_reads_bytes_(TableSize) PVOID SMBiosTables,
+    _In_ ULONG TableSize,
+    _Inout_ PCHAR * Strings)
```

Comment:
So here it may be interesting to add an extra `_In_ Ulong StringArraySize` of sorts, that allows the user to tell the function that the Strings array contains a certain number of elements. This can allow the function to control that the user gives a "correctly"-sized array.

Reply:
Or go with a sal annotation that describes this :)

Reply:
Annotated with `_Inout_updates_(ID_STRINGS_MAX)` as suggested by Mark.

Since I already have upcoming code that depends on this, making significant changes in function arguments is painful. :sweat_smile: